src/operator/controllers/intents/reconciler.go

```diff
@@ -0,0 +1,322 @@
+	return &Reconciler{
+		client:            client,
+		scheme:            scheme,
+		serviceIdResolver: serviceIdResolver,
+		recorder:          eventRecorder,
+	}
+}
+
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(controller.Options{RecoverPanic: lo.ToPtr(true)}).
+		For(&otterizev1alpha3.ClientIntents{}).
+		Watches(&otterizev1alpha3.PostgreSQLServerConfig{}, handler.EnqueueRequestsFromMapFunc(r.mapPGServerConfToClientIntents)).
+		Watches(&v1.Pod{}, handler.EnqueueRequestsFromMapFunc(r.mapPodToClientIntents)).
+		Complete(r)
+}
+
+func (r *Reconciler) mapPodToClientIntents(_ context.Context, obj client.Object) []reconcile.Request {
+	pod := obj.(*v1.Pod)
+	otterizeIdentity, err := r.serviceIdResolver.ResolvePodToServiceIdentity(context.Background(), pod)
+	if err != nil {
+		logrus.Errorf("Failed resovling Otterize identity for pod %s: %v", pod.Name, err)
+	}
+	fullClientName := fmt.Sprintf("%s.%s", otterizeIdentity.Name, otterizeIdentity.Namespace)
+	logrus.Infof("Enqueueing client intents for client '%s'", fullClientName)
+
+	clientIntentsList := otterizev1alpha3.ClientIntentsList{}
+	err = r.client.List(context.Background(),
+		&clientIntentsList,
+		&client.MatchingFields{OtterizeIntentsClientIndexNameField: fullClientName},
+	)
+
+	if err != nil {
+		logrus.Errorf("Failed to list intents for client %s: %v", fullClientName, err)
+	}
+	requests := make([]reconcile.Request, 0)
+	for _, clientIntents := range clientIntentsList.Items {
+		request := reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      clientIntents.Name,
+				Namespace: clientIntents.Namespace,
+			},
+		}
+		requests = append(requests, request)
+	}
+
+	return requests
+}
+
+func (r *Reconciler) mapPGServerConfToClientIntents(ctx context.Context, obj client.Object) []reconcile.Request {
+	pgServerConf := obj.(*otterizev1alpha3.PostgreSQLServerConfig)
+	logrus.Infof("Enqueueing client intents for PostgreSQLServerConfig change %s", pgServerConf.Name)
+
+	intentsToReconcile := make([]otterizev1alpha3.ClientIntents, 0)
+	intentsList := otterizev1alpha3.ClientIntentsList{}
+	dbInstanceName := pgServerConf.Name
+	err := r.client.List(context.Background(),
+		&intentsList,
+		&client.MatchingFields{otterizev1alpha3.OtterizeTargetServerIndexField: dbInstanceName},
+	)
+	if err != nil {
+		logrus.Errorf("Failed to list client intents targeting %s: %v", dbInstanceName, err)
+	}
+	intentsToReconcile = append(intentsToReconcile, intentsList.Items...)
+
+	requests := make([]reconcile.Request, 0)
+	for _, clientIntents := range intentsToReconcile {
+		requests = append(requests, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      clientIntents.Name,
+				Namespace: clientIntents.Namespace,
+			},
+		})
+	}
+	return requests
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
BTW, now that the integration is moved to the OSS, I think achieving separate passwords can be made potentially easier.

Reply:
Since this is on the open-source putting this here: the way the credentials operator works, is it issues a credential to identify the workload, no matter the kind of credential. So, one AWS role, one Azure role, one X.509/SPIFFE credential, and one username. The security model we aim to achieve is zero-trust: that a single workload being compromised will not compromise any other workload's access. This means that the server being compromised should also not compromise any other workloads. For passwords, this means that the server must not keep the password in plaintext or any format that would compromise other servers if the server is compromised. But, this is in the database server's threat model rather than Otterize's: storing passwords in a way that the original password cannot be recovered.